vault/logical_system.go

```diff
@@ -5542,6 +5545,397 @@ func checkListingVisibility(visibility ListingVisibilityType) error {
+	}
+
+	if ok {
+		filters.MessageType(typeValue.(string))
+	}
+
+	activeValue, ok, err := d.GetOkErr("active")
+	if err != nil {
+		return logical.ErrorResponse("invalid active parameter value: %s", err), nil
+	}
+
+	if ok {
+		filters.Active(activeValue.(bool))
+	}
+
+	entries, err := b.Core.uiConfig.ListCustomMessages(ctx, filters)
+	if err != nil {
+		return logical.ErrorResponse("failed to retrieve list of custom messages: %s", err), nil
+	}
+
+	messages := make(map[string]any)
+	keys := make([]string, len(entries))
+
+	for i, entry := range entries {
+		entryMap := make(map[string]any)
+		entryMap["title"] = entry.Title
+		entryMap["type"] = entry.MessageType
+		entryMap["authenticated"] = entry.Authenticated
+		entryMap["start_time"] = entry.StartTime
+		entryMap["end_time"] = entry.EndTime
+		entryMap["active"] = entry.active
+
+		messages[entry.Id] = entryMap
+		keys[i] = entry.Id
+	}
+
+	return logical.ListResponseWithInfo(keys, messages), nil
+}
+
+func (b *SystemBackend) handleCreateCustomMessages(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var err error
+
+	var title string
+	titleValue, ok, err := d.GetOkErr("title")
+	if err != nil {
+		return logical.ErrorResponse("invalid title parameter value: %s", err), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing title parameter value"), nil
+	}
+
+	title = titleValue.(string)
+
+	var authenticated bool
+	authenticatedValue, ok, err := d.GetOkErr("authenticated")
+	if err != nil {
+		return logical.ErrorResponse("invalid authenticated parameter value: %s", err), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing authenticated parameter value"), nil
+	}
+
+	authenticated = authenticatedValue.(bool)
+
+	var messageType string
+	messageTypeValue, ok, err := d.GetOkErr("type")
+	if err != nil {
+		return logical.ErrorResponse("invalid type parameter value: %s", err), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing type parameter value"), nil
+	}
+
+	messageType = messageTypeValue.(string)
+
+	var message string
+	messageValue, ok, err := d.GetOkErr("message")
+	if err != nil {
+		return logical.ErrorResponse("invalid message parameter value: %s", err), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing message parameter value"), nil
+	}
+
+	message = messageValue.(string)
+
+	var startTime time.Time
+	startTimeValue, ok, err := d.GetOkErr("start_time")
+	if err != nil {
+		return logical.ErrorResponse("invalid start_time parameter value"), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing start_time parameter value"), nil
+	}
+
+	startTime = startTimeValue.(time.Time)
+
+	var endTime time.Time
+	endTimeValue, ok, err := d.GetOkErr("end_time")
+	if err != nil {
+		return logical.ErrorResponse("invalid end_time parameter value"), nil
+	}
+
+	if !ok {
+		return logical.ErrorResponse("missing end_time parameter value"), nil
+	}
+
```

Comment:
I think we should have some additional validation here, in particular, I think we should fail if:
- end_time is before now
- start_time is before now (maybe?)
- start_time is after or equal to end time

Reply:
I don't see a problem with having a start_time that is in the past (before Now). That would simply cause the message to be immediately active, which is fine.